config/value.go

```diff
@@ -382,216 +394,263 @@ func convertValue(value interface{}, targetType reflect.Type) (interface{}, erro
+
+		// Iterate until we find first missing value.
+		if v2 := global.Get(arrayKey); v2.HasValue() {
+			val := reflect.New(elementType).Elem()
+
+			// Unmarshal current element.
+			if err := cv.unmarshal(arrayKey, val, ""); err != nil {
+				return err
+			}
+
+			// Append element to the slice
+			if destSlice.Len() <= ai {
+				destSlice = reflect.Append(destSlice, reflect.Zero(elementType))
+			}
+
+			destSlice.Index(ai).Set(val)
+		} else {
+			break
+		}
+	}
+
+	if destSlice.Len() > 0 {
+		fieldValue.Set(destSlice)
+	}
+
+	return nil
+}
+
+// Sets value to a map type.
+func (cv Value) mapping(childKey string, value reflect.Value) error {
+	valueType := value.Type()
+	global := cv.getGlobalProvider()
+
+	val := global.Get(childKey).Value()
+
+	// We fallthrough for interface to maps
+	if valueType.Kind() == reflect.Interface {
+		value.Set(reflect.ValueOf(val))
+		return nil
+	}
+
+	if val == nil {
+		return nil
+	}
+
+	destMap := reflect.ValueOf(reflect.MakeMap(valueType).Interface())
+
+	// child yamlNode parsed from yaml file is of type map[interface{}]interface{}
+	// type casting here makes sure that we are iterating over a parsed map.
+	if v, ok := val.(map[interface{}]interface{}); ok {
+		childKey += _separator
+		for key := range v {
+			mapKey := childKey + fmt.Sprintf("%v", key)
+			itemValue := reflect.New(valueType.Elem()).Elem()
+
+			// Try to unmarshal value and save it in the map.
+			if err := cv.unmarshal(mapKey, itemValue, ""); err != nil {
+				return err
+			}
+
+			destMap.SetMapIndex(reflect.ValueOf(key), itemValue)
+		}
+
+		value.Set(destMap)
+	}
+
+	return nil
+}
+
+// Sets value to an object type.
+func (cv Value) object(childKey string, value reflect.Value) error {
+	valueType := value.Type()
+	global := cv.getGlobalProvider()
+
+	ntt := derefType(valueType)
+	newTarget := reflect.New(ntt)
+	v2 := global.Get(childKey)
+
+	if !v2.HasValue() && valueType.Kind() == reflect.Ptr {
+		// in this case we will keep the pointer value as not defined.
+		return nil
+	}
+
+	if err := v2.PopulateStruct(newTarget.Interface()); err != nil {
+		return errors.Wrap(err, "unable to populate struct of object target")
+	}
+
+	// if the target is not a pointer, deref the value
+	// for copy semantics
+	if valueType.Kind() != reflect.Ptr {
+		newTarget = newTarget.Elem()
+	}
+
+	if value.CanSet() {
+		value.Set(newTarget)
+	}
+
+	return nil
+}
+
+// Walk through the struct and start asking the providers for values at each key.
+//
+// - for individual values, we terminate
+// - for array values, we start asking for indexes
+// - for object values, we recurse.
+func (cv Value) valueStruct(key string, target interface{}) error {
+	tarGet := reflect.Indirect(reflect.ValueOf(target))
+	targetType := tarGet.Type()
 
 	for i := 0; i < targetType.NumField(); i++ {
 		field := targetType.Field(i)
-		fieldType := field.Type
 		fieldName := field.Name
+		if fieldName[0] != strings.ToUpper(fieldName)[0] {
```

Comment:
Why do we need to skip un-exported fields? 
Should this be an error to the user to export the fields, since config should always expect exported fields or shouldn't care at all.